actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -860,6 +863,36 @@ def merge!(other_hash)
       self
     end
 
+    ##
+    # :method: deep_merge
+    # :call-seq: deep_merge(other_hash, &block)
+    #
+    # Returns a new +ActionController::Parameters+ instance with +self+ and +other_hash+ merged recursively.
+    #
+    # Like with +Hash#merge+ in the standard library, a block can be provided
+    # to merge values.
+    #
+    #--
+    # Implemented by ActiveSupport::DeepMergeable#deep_merge.
+
+    ##
+    # :method: deep_merge!
+    # :call-seq: deep_merge!(other_hash, &block)
+    #
+    # Same as +#deep_merge+, but modifies +self+.
+    #
+    #--
+    # Implemented by ActiveSupport::DeepMergeable#deep_merge!.
+
+    def deep_merge!(other_hash)
+      @parameters.deep_merge!(other_hash)
```

Comment:
I note that `merge` and `reverse_merge` both 1) use `other_hash.to_h`, and 2) are not defined as `dup.merge!`

Are either of those points worth special attention?

Reply:
Do we even need to implement this method? Was the point of `DeepMergeable` that the object just need to define `merge!`?

About `dup`, that is a good question. We need to check if it returns the same result as `new_instance_with_inherited_permitted_status`. And about `other_hash.to_h`, if we just delegate to `merge!` we don't need to do anything

Reply:
This diff was able to omit the `#deep_merge` definition, but I wasn't able to omit the `#deep_merge!` implementation while keeping the suite passing.

Reply:
In that case `DeepMergeable` is either broken or should be removed. It was introduced so this class didn't need to implement all the deep merge interface and only implement `merge!` and `deep_merge?`, if we need to implement `merge!`, `deep_merge!` and `deep_merge` why do we need that module?